p11-kit/modules.c

```diff
@@ -185,6 +185,50 @@ const char *p11_config_user_modules = P11_USER_CONFIG_MODULES;
  * P11-KIT FUNCTIONALITY
  */
 
+/**
+ * p11_kit_override_system_files:
+ * @system_conf: the system configuration file (default: system_config_dir/pkcs11.conf)
+ * @user_conf: the user configuration file (default: ~/.config/pkcs11/pkcs11.conf)
+ * @package_modules: location of modules shipped by p11-kit (default: system_config/modules)
+ * @system_modules: location of system pkcs11 modules (default: system_config/modules)
+ * @user_modules: location of user modules (default: ~/.config/pkcs11/modules)
+ *
+ * Overrides the default system configuration files. The
+ * provided values should be accessible for the lifetime
+ * of p11-kit usage.
+ *
+ * When the value %P11_KIT_IGNORE_PTR is provided for any of the locations,
```

Comment:
Can't we simply use NULL, given all the variables are preset to non-NULL strings?

Reply:
What if we want to set some files as being null? (e.g., set the system file and make the user file null)

Reply:
As far as I can see, the code assumes all those variables non-null.  For example:
```p11_config_system_file``` is passed to ```_p11_conf_load_globals```:
https://github.com/p11-glue/p11-kit/blob/master/p11-kit/modules.c#L571
```_p11_conf_load_globals``` calls ```_p11_conf_parse_file```:
https://github.com/p11-glue/p11-kit/blob/master/p11-kit/conf.c#L201
which has a non-NULL assertion on the argument:
https://github.com/p11-glue/p11-kit/blob/master/p11-kit/conf.c#L95

Setting the other 4 variables to NULL also leads to assertion failure.  Maybe you could add some assertions in ```p11_kit_override_system_files``` to fail early?

Reply:
Done. I've also added some text on why P11_KIT_IGNORE_PTR is there instead of NULL.

Reply:
Sorry, that's not what I meant; disregard the last paragraph in my previous comment.  Given that the library doesn't accept NULL as the value of those 5 variables, we could use NULL as an indicator of using a default value, instead of introducing P11_KIT_IGNORE_PTR.

Reply:
I was thinking of future extensibility where null could be accepted, but you are right, no reason to overthink of it. Let's use null. Updated.